File: exchange/ExchangeClient.ts
import WebSocket from "ws";
import ccxt, { Exchange, TradingFees } from "ccxt";
import Order from "./Order.js";
import OrderBook from "./OrderBook.js";
import { Decimal } from "decimal.js";
import RateLimiter from "./RateLimiter.js";
export default class ExchangeClient {
  /*
    Wrapper around ccxt for Binance testnet.
    Handles rate limiting, error handling, and response normalization.
    */
  exchange: Exchange;
  rateLimiter: RateLimiter;
  socket: WebSocket;
  orderBooks: Map<string, OrderBook>;
  private constructor(config: {
    apiKey: string;
    secret: string;
    sandbox: boolean;
    options: any;
    enableRateLimit: boolean;
    wsUrl: string;
  }) {
    /*
        Initialize with config dict containing apiKey, secret, sandbox flag.
        */
    const exchangeId = "binance";
    const exchangeClass = ccxt[exchangeId];
    this.exchange = new exchangeClass({
      apiKey: config.apiKey,
      secret: config.secret,
      sandbox: config.sandbox,
      options: config.options,
      enableRateLimit: config.enableRateLimit,
    });
    this.rateLimiter = new RateLimiter();
    this.orderBooks = new Map();
    this.socket = new WebSocket(config.wsUrl);
    this.socket.addEventListener("message", (event: any) => {
      const data = JSON.parse(event.data);
      if (data.e === "depthUpdate") {
        const symbol = data.s.toString();
        const book = new OrderBook(
          symbol,
          Date.now(),
          data.b, // bids
          data.a, // asks
        );

        this.orderBooks.set(symbol, book);
      }
    });

    this.socket.addEventListener("error", (err) => {
      console.error("WebSocket error", err);
    });
  }

  static async fromConfig(config: {
    apiKey: string;
    secret: string;
    sandbox: boolean;
    options: any;
    enableRateLimit: boolean;
    wsUrl: string;
  }): Promise<ExchangeClient> {
    const client = new ExchangeClient(config);
    try {
      await client.exchange.fetchTime();
      return client;
    } catch (error) {
      throw error;
    }
  }

  async callAPI<T>(name: string, fn: () => Promise<T>): Promise<T> {
    if (!this.rateLimiter.canRequest()) {
      throw "Request blocked";
    }
    let weight = 0;
    const start = Date.now();
    try {
      const res = await fn();
      weight = Number(
        this.exchange.last_response_headers["X-Mbx-Used-Weight-1m"],
      );
      console.log(
        `[EXCHANGE] ${name} ok (${Date.now() - start}ms) weight: ${weight}`,
      );

      return res;
    } catch (err: any) {
      console.error(`[EXCHANGE] ${name} failed`, err?.message);

      if (
        err instanceof ccxt.NetworkError ||
        err instanceof ccxt.ExchangeNotAvailable
      ) {
        throw new Error(`Network error: ${err.message}`);
      }

      if (err instanceof ccxt.AuthenticationError) {
        throw new Error(`Auth error: ${err.message}`);
      }

      if (
        err instanceof ccxt.RateLimitExceeded ||
        err?.message?.includes("429")
      ) {
        console.log("RateLimitExceeded");
      }
      throw err;
    } finally {
      this.rateLimiter.record(weight);
    }
  }

  async fetchOrderBook(
    symbol: string, // "ETH/USDT"
    limit: number = 20, // Number of price levels
  ): Promise<OrderBook | undefined> {
    /*
        Fetch L2 order book snapshot.
        */

    //from ccxt
    // const cctxOrderBook = await this.callAPI("fetchOrderBook", () =>
    //   this.exchange.fetchOrderBook(symbol, limit),
    // );
    //return new OrderBook(
    //   symbol,
    //   Number(cctxOrderBook.timestamp),
    //   cctxOrderBook.bids,
    //   cctxOrderBook.asks,
    // );

    const nsymbol = symbol.replace("/", "");
    const orderBook = this.orderBooks.get(nsymbol);

    return orderBook;
  }
  async fetchBalance(): Promise<{
    [id: string]: { free: Decimal; used: Decimal };
  }> {
    /*
        Fetch account balances.

        Returns:
        {
            'ETH':  {'free': Decimal('10.5'), 'locked': Decimal('0'), 'total': Decimal('10.5')},
            'USDT': {'free': Decimal('20000'), 'locked': Decimal('500'), 'total': Decimal('20500')},
            
        }

        Must filter out zero-balance assets.
        */
    const balances = await this.callAPI("fetchBalance", () =>
      this.exchange.fetchBalance(),
    );
    const nonZero: {
      [id: string]: { free: Decimal; used: Decimal };
    } = {};
    const keys: string[] = Object.keys(balances);
    for (let i = 0; i < keys.length; i++) {
      const balance = balances[keys[i]];
      if (!balance) continue;

      if (balance.total === 0 || isNaN(Number(balance.total))) continue;

      nonZero[keys[i]] = {
        free: new Decimal(Number(balance.free)),
        used: new Decimal(Number(balance.used)),
      };
    }
    return nonZero;
  }

  async createLimitIocOrder(
    symbol: string, // "ETH/USDT"
    side: string, // "buy" or "sell"
    amount: Decimal, // Quantity of base asset
    price: Decimal, // Limit price
  ): Promise<Order> {
    /*
        Place a LIMIT IOC (Immediate Or Cancel) order.

        Returns normalized order result:
        {
            'id': str,
            'symbol': str,
            'side': str,
            'type': 'limit',
            'time_in_force': 'IOC',
            'amount_requested': Decimal,
            'amount_filled': Decimal,
            'avg_fill_price': Decimal,
            'fee': Decimal,
            'fee_asset': str,
            'status': str,  # 'filled', 'partially_filled', 'expired'
            'timestamp': int,
        }

        Must handle: partial fills, rejection, and exchange errors.
        */
    const cctxOrder = await this.callAPI("createLimitOrder", () =>
      this.exchange.createLimitOrder(
        symbol,
        side,
        amount.toNumber(),
        price.toNumber(),
        {
          timeInForce: "IOC",
        },
      ),
    );
    const order = new Order(cctxOrder);
    return order;
  }
  async createMarketOrder(
    symbol: string,
    side: string,
    amount: Decimal,
  ): Promise<Order> {
    /*
        Place a market order. Same return format as create_limit_ioc_order.
        Use sparingly — LIMIT IOC is preferred for arb.
        */
    const cctxOrder = await this.callAPI("createMarketOrder", () =>
      this.exchange.createMarketOrder(symbol, side, amount.toNumber()),
    );
    const order = new Order(cctxOrder);
    return order;
  }
  async cancelOrder(orderId: string, symbol: string): Promise<Order> {
    // Cancel an open order. Returns order status after cancel."""
    const cctxOrder = await this.callAPI("cancelOrder", () =>
      this.exchange.cancelOrder(orderId, symbol),
    );
    const order = new Order(cctxOrder);
    return order;
  }
  async fetchOrderStatus(orderId: string, symbol: string): Promise<string> {
    // Check current status of an order."""
    const orderStatus = await this.callAPI("fetchOrderStatus", () =>
      this.exchange.fetchOrderStatus(orderId, symbol),
    );
    return orderStatus;
  }
  async getTradingFees(symbol: string): Promise<TradingFees | undefined> {
    /*
        Returns fee structure:
        {'maker': Decimal('0.001'), 'taker': Decimal('0.001')}
        */
    const fees = await this.callAPI("fetchTradingFees", () =>
      this.exchange.fetchTradingFees(symbol),
    );
    return fees;
  }
}

File: exchange/exchange_client.py
# exchange/client.py

import asyncio
import json
import time
from typing import Awaitable, Callable, TypeVar
import ccxt
import websockets
from decimal import Decimal

from exchange.order import Order
from exchange.order_book import OrderBook
from exchange.rate_limiter import RateLimiter

T = TypeVar("T")


class ExchangeClient:
    """
    Wrapper around ccxt for Binance testnet.
    Handles rate limiting, error handling, and response normalization.
    """

    def __init__(self, config: dict):
        """
        Initialize with config dict containing apiKey, secret, sandbox flag.
        Must validate connection on init (fetch server time or status).
        """
        exchangeClass = ccxt.binance
        self.exchange = exchangeClass(
            {
                "apiKey": config.get("apiKey", ""),
                "secret": config.get(
                    "secret",
                ),
                "sandbox": config.get("sandbox", True),
                "options": config.get("options", {}),
                "enableRateLimit": config.get("enableRateLimit", True),
            }
        )

        self.rateLimiter = RateLimiter()
        self.order_books: dict[str, OrderBook] = dict()

    async def start(self):
        asyncio.create_task(self.watch_order_book())

    async def watch_order_book(self):
        async with websockets.connect(
            "wss://stream.binance.com:9443/ws/ethusdt@depth"
        ) as ws:
            async for message in ws:
                print(message)
                data = await ws.recv()
                data_json = json.loads(data)
                symbol = data_json["s"]
                book = OrderBook(
                    symbol,
                    (int)(time.time()),
                    data_json["b"],  # bids
                    data_json["a"],  # asks
                )
                self.order_books[symbol] = book

    def call_API(self, name: str, fn: Callable) -> T:
        if not self.rateLimiter.can_request():
            raise RuntimeError("Request blocked")

        weight = 0
        start = time.time()

        try:
            res = fn()
            weight = int(
                self.exchange.last_response_headers.get("X-Mbx-Used-Weight-1m", 0)
            )
            print(
                f"[EXCHANGE] {name} ok ({(time.time() - start) * 1000:.0f}ms) weight: {weight}"
            )
            return res

        except Exception as err:
            print(f"[EXCHANGE] {name} failed: {err}")

            if isinstance(err, (ccxt.NetworkError, ccxt.ExchangeNotAvailable)):
                raise ConnectionError(f"Network error: {err}") from err

            if isinstance(err, ccxt.AuthenticationError):
                raise PermissionError(f"Auth error: {err}") from err

            if isinstance(err, ccxt.RateLimitExceeded) or "429" in str(err):
                print("RateLimitExceeded")

            raise

        finally:
            self.rateLimiter.record(weight)

    def fetch_order_book_rest(self, symbol: str, depth: int = 20):
        result = self.call_API(
            "fetchOrderBook",
            lambda: self.exchange.fetch_order_book(symbol=symbol, limit=depth),
        )
        ob = OrderBook(
            symbol=symbol,
            timestamp=result.get("timestamp"),
            bids=result.get("bids"),
            asks=result.get("asks"),
        )
        return ob

    def fetch_order_book(
        self, symbol: str, limit: int = 20  # "ETH/USDT"  # Number of price levels
    ) -> OrderBook:
        """
        Fetch L2 order book snapshot.

        Returns normalized dict:
        {
            'symbol': 'ETH/USDT',
            'timestamp': 1706000000000,
            'bids': [(price, qty), ...],  # Sorted best→worst
            'asks': [(price, qty), ...],  # Sorted best→worst
            'best_bid': (price, qty),
            'best_ask': (price, qty),
            'mid_price': Decimal,
            'spread_bps': Decimal,
        }
        """
        order_book: OrderBook = self.order_books.get(symbol)

        if order_book is None:
            return None
        return_order_book = OrderBook(
            symbol,
            order_book.timestamp,
            order_book.bids[:limit],
            order_book.asks[:limit],
        )

        return return_order_book

    def fetch_balance(self) -> dict[str, dict]:
        """
        Fetch account balances.

        Returns:
        {
            'ETH':  {'free': Decimal('10.5'), 'locked': Decimal('0'), 'total': Decimal('10.5')},
            'USDT': {'free': Decimal('20000'), 'locked': Decimal('500'), 'total': Decimal('20500')},
            ...
        }

        Must filter out zero-balance assets.
        """
        result = self.call_API(
            "fetchBalance",
            self.exchange.fetchBalance,
        )
        info = result.get("info")
        balances = info.get("balances", [])
        nonZero: dict[str, dict] = dict()
        for balance in balances:
            if not balance:
                continue

            if balance.get("free", 0) + balance.get("locked", 0) == 0:
                continue
            nonZero[balance.get("asset")] = {
                "free": Decimal(balance["free"]),
                "used": Decimal(balance["locked"]),
            }

        return nonZero

    def create_limit_ioc_order(
        self,
        symbol: str,  # "ETH/USDT"
        side: str,  # "buy" or "sell"
        amount: float,  # Quantity of base asset
        price: float,  # Limit price
    ) -> dict:
        """
        Place a LIMIT IOC (Immediate Or Cancel) order.

        Returns normalized order result:
        {
            'id': str,
            'symbol': str,
            'side': str,
            'type': 'limit',
            'time_in_force': 'IOC',
            'amount_requested': Decimal,
            'amount_filled': Decimal,
            'avg_fill_price': Decimal,
            'fee': Decimal,
            'fee_asset': str,
            'status': str,  # 'filled', 'partially_filled', 'expired'
            'timestamp': int,
        }

        Must handle: partial fills, rejection, and exchange errors.
        """
        cctx_order = self.call_API(
            "createLimitOrder",
            lambda: self.exchange.create_limit_order(
                symbol,
                side,
                amount,
                price,
                {
                    "timeInForce": "IOC",
                },
            ),
        )
        order = Order(cctx_order)
        return order

    def create_market_order(
        self,
        symbol: str,
        side: str,
        amount: float,
    ) -> dict:
        """
        Place a market order. Same return format as create_limit_ioc_order.
        Use sparingly — LIMIT IOC is preferred for arb.
        """
        cctx_order = self.call_API(
            "createMarketOrder",
            lambda: self.exchange.create_market_order(symbol, side, amount),
        )
        order = Order(cctx_order)
        return order

    def cancel_order(self, order_id: str, symbol: str) -> dict:
        """Cancel an open order. Returns order status after cancel."""
        cctx_order = self.call_API(
            "cancelOrder",
            lambda: self.exchange.cancel_order(order_id, symbol),
        )
        order = Order(cctx_order)
        return order

    def fetch_order_status(self, order_id: str, symbol: str) -> dict:
        """Check current status of an order."""
        order_status = self.call_API(
            "fetchOrderStatus",
            lambda: self.exchange.fetch_order_status(order_id, symbol),
        )
        return order_status

    def get_trading_fees(self, symbol: str) -> dict:
        """
        Returns fee structure:
        {'maker': Decimal('0.001'), 'taker': Decimal('0.001')}
        """
        fees = self.call_API(
            "fetchTradingFees",
            lambda: self.exchange.fetch_trading_fees(symbol),
        )
        return fees
